lib/src/backends/record_replay/recording_file.dart

```diff
@@ -89,3 +138,151 @@ class RecordingFile extends RecordingFileSystemEntity<File, io.File>
+        super(stream);
+
+  @override
+  void onData(List<int> event) {
+    if (_sink == null) {
+      _sink = file.openWrite();
+    }
+    _sink.add(event);
+  }
+
+  @override
+  void onDone() {
+    if (_sink != null) {
+      _sink.close();
+    }
+  }
+
+  @override
+  dynamic get serializedValue => '!${file.basename}';
+
+  // TODO(tvolkert): remove `.then()` once Dart 1.22 is in stable
+  @override
+  Future<Null> get complete =>
+      Future.wait(<Future<dynamic>>[super.complete, _sink.done]).then((_) {});
+}
+
+class _ByteArrayFutureReference extends FutureReference<List<int>> {
+  final File file;
+
+  _ByteArrayFutureReference(
+      MutableRecording recording, String name, Future<List<int>> future)
+      : file = recording.newFile(name),
+        super(future);
+
+  @override
+  Future<List<int>> get value => super.value.then((List<int> bytes) async {
+        await file.writeAsBytes(bytes, flush: true);
+        return bytes;
+      });
+
+  @override
+  dynamic get serializedValue => '!${file.basename}';
+}
+
+class _ByteArrayReference extends ResultReference<List<int>> {
+  final File file;
+  final List<int> bytes;
+
+  _ByteArrayReference(MutableRecording recording, String name, this.bytes)
+      : file = recording.newFile(name);
+
+  @override
+  List<int> get value {
+    file.writeAsBytesSync(bytes, flush: true);
+    return bytes;
+  }
+
+  @override
+  List<int> get recordedValue => bytes;
+
+  @override
+  dynamic get serializedValue => '!${file.basename}';
+}
+
+class _FileContentFutureReference extends FutureReference<String> {
+  final File file;
+
+  _FileContentFutureReference(
+      MutableRecording recording, String name, Future<String> future)
+      : file = recording.newFile(name),
+        super(future);
+
+  @override
+  Future<String> get value => super.value.then((String content) async {
+        await file.writeAsString(content, flush: true);
+        return content;
+      });
+
+  @override
+  dynamic get serializedValue => '!${file.basename}';
+}
+
+class _FileContentReference extends ResultReference<String> {
+  final File file;
+  final String content;
+
+  _FileContentReference(MutableRecording recording, String name, this.content)
+      : file = recording.newFile(name);
+
+  @override
+  String get value {
+    file.writeAsStringSync(content, flush: true);
+    return content;
+  }
+
+  @override
+  String get recordedValue => content;
+
+  @override
+  dynamic get serializedValue => '!${file.basename}';
+}
+
+class _LinesFutureReference extends FutureReference<List<String>> {
+  final File file;
+
+  _LinesFutureReference(
+      MutableRecording recording, String name, Future<List<String>> future)
+      : file = recording.newFile(name),
+        super(future);
+
+  @override
+  Future<List<String>> get value =>
+      super.value.then((List<String> lines) async {
+        await file.writeAsString(lines.join('\n'), flush: true);
+        return lines;
+      });
```

Comment:
Totally optional readability suggestion: I'd be tempted to use a block rather than a mix of top-level expression and a then implemented as a block.